File: client/src/main/java/com/msc/serverbrowser/gui/views/MainView.java
package com.msc.serverbrowser.gui.views;

import com.msc.serverbrowser.Client;
import com.msc.serverbrowser.gui.View;
import com.msc.serverbrowser.util.windows.OSUtility;

import javafx.beans.property.DoubleProperty;
import javafx.event.ActionEvent;
import javafx.event.EventHandler;
import javafx.scene.Node;
import javafx.scene.Parent;
import javafx.scene.control.Hyperlink;
import javafx.scene.control.Label;
import javafx.scene.control.ProgressBar;
import javafx.scene.control.ScrollPane;
import javafx.scene.control.ToggleButton;
import javafx.scene.control.ToggleGroup;
import javafx.scene.control.Tooltip;
import javafx.scene.layout.HBox;
import javafx.scene.layout.Priority;
import javafx.scene.layout.VBox;

/**
 * Class containing the component hierarchy for the main view.
 * <p>
 * It contains the menu bar, the active view and the bottom bar.
 * </p>
 *
 * @author Marcel
 * @since 10.01.2018
 */
public class MainView {
	private final HBox rootPane;

	private final ToggleButton	menuItemServers;
	private final ToggleButton	menuItemUser;
	private final ToggleButton	menuItemVersion;
	private final ToggleButton	menuItemFiles;
	private final ToggleButton	menuItemSettings;

	private final ScrollPane contentScrollPane;

	private final Hyperlink	githubLink;
	private final Hyperlink	helpLink;
	private final Hyperlink	donateLink;

	private final HBox bottomBarCustom;

	private final Label			globalProgressLabel;
	private final ProgressBar	globalProgressBar;

	/**
	 * Initializes the whole view.
	 */
	public MainView() {
		rootPane = new HBox();
		rootPane.setPrefSize(800, 500);
		rootPane.getStyleClass().add("root-pane");

		final VBox menuContainer = new VBox();
		menuContainer.getStyleClass().add("tabPane");

		final ToggleGroup menuItemToggleGroup = new ToggleGroup();
		final String menuItemStyleClass = "MenuItem";

		menuItemServers = new ToggleButton("\uf0c9");
		menuItemServers.getStyleClass().add(menuItemStyleClass);
		menuItemUser = new ToggleButton("\uf007");
		menuItemUser.getStyleClass().add(menuItemStyleClass);
		menuItemVersion = new ToggleButton("\uf0ed");
		menuItemVersion.getStyleClass().add(menuItemStyleClass);
		menuItemFiles = new ToggleButton("\uf07b");
		menuItemFiles.getStyleClass().add(menuItemStyleClass);
		menuItemSettings = new ToggleButton("\uf013");
		menuItemSettings.getStyleClass().add(menuItemStyleClass);

		menuItemToggleGroup.getToggles().addAll(menuItemServers, menuItemUser, menuItemVersion, menuItemFiles, menuItemSettings);
		menuContainer.getChildren().addAll(menuItemServers, menuItemUser, menuItemVersion, menuItemFiles, menuItemSettings);

		final ScrollPane menuScrollPane = new ScrollPane(menuContainer);
		menuScrollPane.setFitToHeight(true);
		menuScrollPane.setFitToWidth(true);
		menuScrollPane.getStyleClass().add("tabScrollPane");

		final VBox mainContentPane = new VBox();
		HBox.setHgrow(mainContentPane, Priority.ALWAYS);
		contentScrollPane = new ScrollPane();
		contentScrollPane.setFitToHeight(true);
		contentScrollPane.setFitToWidth(true);
		contentScrollPane.getStyleClass().add("viewContent");
		VBox.setVgrow(contentScrollPane, Priority.ALWAYS);

		final HBox bottomBar = new HBox();
		bottomBar.getStyleClass().add("bottom-bar");

		githubLink = new Hyperlink("\uf09b");
		githubLink.getStyleClass().add("info-icon");
		githubLink.setTooltip(new Tooltip(Client.getString("openGithubTooltip")));
		githubLink.setFocusTraversable(false);
		helpLink = new Hyperlink("\uf059");
		helpLink.getStyleClass().add("info-icon");
		helpLink.setTooltip(new Tooltip(Client.getString("openGithubWikiTooltip")));
		helpLink.setFocusTraversable(false);
		donateLink = new Hyperlink(Client.getString("donate") + " \uf0d6");
		donateLink.getStyleClass().add("donate-button");
		donateLink.setTooltip(new Tooltip(Client.getString("openDonationPageTooltip")));
		donateLink.setMaxHeight(Double.MAX_VALUE);
		donateLink.setFocusTraversable(false);

		bottomBarCustom = new HBox();
		bottomBarCustom.getStyleClass().add("bottom-bar-custom");
		HBox.setHgrow(bottomBarCustom, Priority.ALWAYS);

		final HBox progressBarContainer = new HBox();
		progressBarContainer.getStyleClass().add("global-progress-bar-container");
		globalProgressLabel = new Label();
		globalProgressBar = new ProgressBar(0.0);
		progressBarContainer.getChildren().addAll(globalProgressLabel, globalProgressBar);

		bottomBar.getChildren().addAll(githubLink, helpLink, donateLink, bottomBarCustom, progressBarContainer);

		mainContentPane.getChildren().add(contentScrollPane);
		mainContentPane.getChildren().add(bottomBar);

		rootPane.getChildren().add(menuScrollPane);
		rootPane.getChildren().add(mainContentPane);
	}

	/**
	 * Inserts the {@link Node} into the {@link #contentScrollPane}.
	 *
	 * @param node the {@link Node} to be inserted into the {@link #contentScrollPane}
	 */
	public void setActiveViewNode(final Node node) {
		contentScrollPane.setContent(node);
	}

	/**
	 * Adds {@link Node}s to the custom part of the BottomBar
	 *
	 * @param nodes the {@link Node}s to be added
	 */
	public void addToBottomBar(final Node... nodes) {
		bottomBarCustom.getChildren().addAll(nodes);
	}

	/**
	 * Removes all {@link Node}s that have been added to the {@link #bottomBarCustom}.
	 */
	public void removeNodesFromBottomBar() {
		bottomBarCustom.getChildren().clear();
	}

	/**
	 * Sets the text for the global {@link ProgressBar}.
	 *
	 * @param text the text to be set
	 */
	public void setGlobalProgressBarText(final String text) {
		globalProgressLabel.setText(text);
	}

	/**
	 * @return the {@link DoubleProperty ProgressProperty} for the {@link #globalProgressBar}
	 */
	public DoubleProperty globalProgressProperty() {
		return globalProgressBar.progressProperty();
	}

	/**
	 * Selects the proper menu item, depending on which {@link View} was given.
	 *
	 * @param view the {@link View} to select the menu item for
	 */
	public void selectMenuItemForView(final View view) {
		switch (view) {
			case SERVERS:
				menuItemServers.setSelected(true);
				break;
			case USERNAME_CHANGER:
				menuItemUser.setSelected(true);
				break;
			case VERSION_CHANGER:
				menuItemVersion.setSelected(true);
				break;
			case FILES:
				menuItemFiles.setSelected(true);
				break;
			case SETTINGS:
				menuItemSettings.setSelected(true);
				break;
			default:
				throw new IllegalArgumentException("This View hasn't been implemented or is invalid: " + view);
		}
	}

	/**
	 * Sets the {@link EventHandler} to handle all {@link ActionEvent}s on the
	 * {@link #menuItemServers}.
	 *
	 * @param handler {@link EventHandler} to be set
	 */
	public void setMenuItemAllAction(final EventHandler<ActionEvent> handler) {
		menuItemServers.setOnAction(handler);
	}

	/**
	 * Sets the {@link EventHandler} to handle all {@link ActionEvent}s on the
	 * {@link #menuItemUser}.
	 *
	 * @param handler {@link EventHandler} to be set
	 */
	public void setMenuItemUsernameAction(final EventHandler<ActionEvent> handler) {
		menuItemUser.setOnAction(handler);
	}

	/**
	 * Sets the {@link EventHandler} to handle all {@link ActionEvent}s on the
	 * {@link #menuItemVersion}.
	 *
	 * @param handler {@link EventHandler} to be set
	 */
	public void setMenuItemVersionAction(final EventHandler<ActionEvent> handler) {
		menuItemVersion.setOnAction(handler);
	}

	/**
	 * Sets the {@link EventHandler} to handle all {@link ActionEvent}s on the
	 * {@link #menuItemFiles}.
	 *
	 * @param handler {@link EventHandler} to be set
	 */
	public void setMenuItemFilesAction(final EventHandler<ActionEvent> handler) {
		menuItemFiles.setOnAction(handler);
	}

	/**
	 * Sets the {@link EventHandler} to handle all {@link ActionEvent}s on the
	 * {@link #menuItemSettings}.
	 *
	 * @param handler {@link EventHandler} to be set
	 */
	public void setMenuItemSettingsAction(final EventHandler<ActionEvent> handler) {
		menuItemSettings.setOnAction(handler);
	}

	/**
	 * @return {@link #rootPane}
	 */
	public Parent getRootPane() {
		return rootPane;
	}

	/**
	 * Sets a browse action for the {@link #githubLink} using the given {@link String} as the URL.
	 *
	 * @param string the URL
	 */
	public void setGitHubHyperlink(final String string) {
		githubLink.setOnAction(__ -> OSUtility.browse(string));
	}

	/**
	 * Sets a browse action for the {@link #helpLink} using the given {@link String} as the URL.
	 *
	 * @param string the URL
	 */
	public void setHelpHyperlink(final String string) {
		helpLink.setOnAction(__ -> OSUtility.browse(string));
	}

	/**
	 * Sets a browse action for the {@link #donateLink} using the given {@link String} as the URL.
	 *
	 * @param string the URL
	 */
	public void setDonateHyperlink(final String string) {
		donateLink.setOnAction(__ -> OSUtility.browse(string));
	}
}

File: src/main/kotlin/com/msc/serverbrowser/gui/views/MainView.kt
package com.msc.serverbrowser.gui.views

import javafx.beans.property.DoubleProperty
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.scene.Node
import javafx.scene.control.*
import javafx.scene.layout.HBox
import javafx.scene.layout.Priority
import javafx.scene.layout.VBox
import com.msc.serverbrowser.Client
import com.msc.serverbrowser.gui.View
import com.msc.serverbrowser.util.windows.OSUtility

/**
 * Class containing the component hierarchy for the main view.
 *
 *
 * It contains the menu bar, the active view and the bottom bar.
 *
 *
 * @author Marcel
 * @since 10.01.2018
 */
class MainView {
    val rootPane: HBox = HBox()

    private val menuItemServers: ToggleButton
    private val menuItemUser: ToggleButton
    private val menuItemVersion: ToggleButton
    private val menuItemFiles: ToggleButton
    private val menuItemSettings: ToggleButton

    private val contentScrollPane: ScrollPane

    private val githubLink: Hyperlink
    private val helpLink: Hyperlink
    private val donateLink: Hyperlink

    private val bottomBarCustom: HBox

    private val globalProgressLabel: Label
    private val globalProgressBar: ProgressBar

    /**
     * Initializes the whole view.
     */
    init {
        rootPane.setPrefSize(800.0, 500.0)
        rootPane.styleClass.add("root-pane")

        val menuContainer = VBox()
        menuContainer.styleClass.add("tabPane")

        val menuItemToggleGroup = ToggleGroup()
        val menuItemStyleClass = "MenuItem"

        menuItemServers = ToggleButton("\uf0c9")
        menuItemServers.styleClass.add(menuItemStyleClass)
        menuItemUser = ToggleButton("\uf007")
        menuItemUser.styleClass.add(menuItemStyleClass)
        menuItemVersion = ToggleButton("\uf0ed")
        menuItemVersion.styleClass.add(menuItemStyleClass)
        menuItemFiles = ToggleButton("\uf07b")
        menuItemFiles.styleClass.add(menuItemStyleClass)
        menuItemSettings = ToggleButton("\uf013")
        menuItemSettings.styleClass.add(menuItemStyleClass)

        menuItemToggleGroup.toggles.addAll(menuItemServers, menuItemUser, menuItemVersion, menuItemFiles, menuItemSettings)
        menuContainer.children.addAll(menuItemServers, menuItemUser, menuItemVersion, menuItemFiles, menuItemSettings)

        val menuScrollPane = ScrollPane(menuContainer)
        menuScrollPane.isFitToHeight = true
        menuScrollPane.isFitToWidth = true
        menuScrollPane.styleClass.add("tabScrollPane")

        val mainContentPane = VBox()
        HBox.setHgrow(mainContentPane, Priority.ALWAYS)
        contentScrollPane = ScrollPane()
        contentScrollPane.isFitToHeight = true
        contentScrollPane.isFitToWidth = true
        contentScrollPane.styleClass.add("viewContent")
        VBox.setVgrow(contentScrollPane, Priority.ALWAYS)

        val bottomBar = HBox()
        bottomBar.styleClass.add("bottom-bar")

        githubLink = Hyperlink("\uf09b")
        githubLink.styleClass.add("info-icon")
        githubLink.tooltip = Tooltip(Client.getString("openGithubTooltip"))
        githubLink.isFocusTraversable = false
        helpLink = Hyperlink("\uf059")
        helpLink.styleClass.add("info-icon")
        helpLink.tooltip = Tooltip(Client.getString("openGithubWikiTooltip"))
        helpLink.isFocusTraversable = false
        donateLink = Hyperlink(Client.getString("donate") + " \uf0d6")
        donateLink.styleClass.add("donate-button")
        donateLink.tooltip = Tooltip(Client.getString("openDonationPageTooltip"))
        donateLink.maxHeight = java.lang.Double.MAX_VALUE
        donateLink.isFocusTraversable = false

        bottomBarCustom = HBox()
        bottomBarCustom.styleClass.add("bottom-bar-isCustom")
        HBox.setHgrow(bottomBarCustom, Priority.ALWAYS)

        val progressBarContainer = HBox()
        progressBarContainer.styleClass.add("global-progress-bar-container")
        globalProgressLabel = Label()
        globalProgressBar = ProgressBar(0.0)
        progressBarContainer.children.addAll(globalProgressLabel, globalProgressBar)

        bottomBar.children.addAll(githubLink, helpLink, donateLink, bottomBarCustom, progressBarContainer)

        mainContentPane.children.add(contentScrollPane)
        mainContentPane.children.add(bottomBar)

        rootPane.children.add(menuScrollPane)
        rootPane.children.add(mainContentPane)
    }

    /**
     * Inserts the [Node] into the [.contentScrollPane].
     *
     * @param node the [Node] to be inserted into the [.contentScrollPane]
     */
    fun setActiveViewNode(node: Node) {
        contentScrollPane.content = node
    }

    /**
     * Adds [Node]s to the isCustom part of the BottomBar
     *
     * @param nodes the [Node]s to be added
     */
    fun addToBottomBar(vararg nodes: Node) {
        bottomBarCustom.children.addAll(*nodes)
    }

    /**
     * Removes all [Node]s that have been added to the [.bottomBarCustom].
     */
    fun removeNodesFromBottomBar() {
        bottomBarCustom.children.clear()
    }

    /**
     * Sets the text for the global [ProgressBar].
     *
     * @param text the text to be set
     */
    fun setGlobalProgressBarText(text: String) {
        globalProgressLabel.text = text
    }

    /**
     * @return the [ProgressProperty][DoubleProperty] for the [.globalProgressBar]
     */
    fun globalProgressProperty(): DoubleProperty {
        return globalProgressBar.progressProperty()
    }

    /**
     * Selects the proper menu item, depending on which [View] was given.
     *
     * @param view the [View] to select the menu item for
     */
    fun selectMenuItemForView(view: View) {
        when (view) {
            View.SERVERS -> menuItemServers.isSelected = true
            View.USERNAME_CHANGER -> menuItemUser.isSelected = true
            View.VERSION_CHANGER -> menuItemVersion.isSelected = true
            View.FILES -> menuItemFiles.isSelected = true
            View.SETTINGS -> menuItemSettings.isSelected = true
            else -> throw IllegalArgumentException("This View hasn't been implemented or is invalid: $view")
        }
    }

    /**
     * Sets the [EventHandler] to handle all [ActionEvent]s on the
     * [.menuItemServers].
     *
     * @param handler [EventHandler] to be set
     */
    fun setMenuItemAllAction(handler: EventHandler<ActionEvent>) {
        menuItemServers.onAction = handler
    }

    /**
     * Sets the [EventHandler] to handle all [ActionEvent]s on the
     * [.menuItemUser].
     *
     * @param handler [EventHandler] to be set
     */
    fun setMenuItemUsernameAction(handler: EventHandler<ActionEvent>) {
        menuItemUser.onAction = handler
    }

    /**
     * Sets the [EventHandler] to handle all [ActionEvent]s on the
     * [.menuItemVersion].
     *
     * @param handler [EventHandler] to be set
     */
    fun setMenuItemVersionAction(handler: EventHandler<ActionEvent>) {
        menuItemVersion.onAction = handler
    }

    /**
     * Sets the [EventHandler] to handle all [ActionEvent]s on the
     * [.menuItemFiles].
     *
     * @param handler [EventHandler] to be set
     */
    fun setMenuItemFilesAction(handler: EventHandler<ActionEvent>) {
        menuItemFiles.onAction = handler
    }

    /**
     * Sets the [EventHandler] to handle all [ActionEvent]s on the
     * [.menuItemSettings].
     *
     * @param handler [EventHandler] to be set
     */
    fun setMenuItemSettingsAction(handler: EventHandler<ActionEvent>) {
        menuItemSettings.onAction = handler
    }

    /**
     * Sets a browse action for the [.githubLink] using the given [String] as the URL.
     *
     * @param string the URL
     */
    fun setGitHubHyperlink(string: String) {
        githubLink.setOnAction { OSUtility.browse(string) }
    }

    /**
     * Sets a browse action for the [.helpLink] using the given [String] as the URL.
     *
     * @param string the URL
     */
    fun setHelpHyperlink(string: String) {
        helpLink.setOnAction { OSUtility.browse(string) }
    }

    /**
     * Sets a browse action for the [.donateLink] using the given [String] as the URL.
     *
     * @param string the URL
     */
    fun setDonateHyperlink(string: String) {
        donateLink.setOnAction { OSUtility.browse(string) }
    }
}
